monitors/test/ContractMonitor.js

```diff
@@ -0,0 +1,295 @@
+
+    // Withdraw from liquidation to settle the dispute event.
+    const txObject1 = await emp.withdrawLiquidation("0", sponsor1, { from: liquidator });
+    await eventClient.clearState();
+
+    // Update the eventClient and check it has the dispute event stored correctly
+    await eventClient._update();
+
+    await contractMonitor.checkForNewDisputeSettlementEvents(() => toWei("1"));
+
+    assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/address/${liquidator}`));
+    assert.isTrue(lastSpyLogIncludes(spy, "(Monitored liquidator bot)"));
+    assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/address/${disputer}`));
+    assert.isTrue(lastSpyLogIncludes(spy, "(Monitored dispute bot)"));
+    assert.isTrue(lastSpyLogIncludes(spy, "failed")); // the disputed was not successful based on settlement price
+    assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/tx/${txObject1.tx}`));
+
+    // Create a second liquidation from a non-monitored address (sponsor1).
+    liquidationTime = (await emp.getCurrentTime()).toNumber();
+    await emp.createLiquidation(sponsor2, { rawValue: toWei("99999") }, { rawValue: toWei("100") }, { from: sponsor1 });
+
+    // Dispute the liquidator from a non-monitor address (sponsor2)
+    await emp.dispute("0", sponsor2, {
+      from: sponsor2
+    });
+
+    // Advance time and settle
+    timeAfterLiquidationLiveness = liquidationTime + 10;
+    await mockOracle.setCurrentTime(timeAfterLiquidationLiveness.toString());
+    await emp.setCurrentTime(timeAfterLiquidationLiveness.toString());
+
+    // Push a price such that the dispute succeeds and ensure the resolution reports correctly. Sponsor2 has 45 units of
+    // debt and 175 units of collateral. price of 2.0: 175 / (45 * 2) = 194% => sufficiently collateralized
+    disputePrice = toWei("2.0");
+    await mockOracle.pushPrice(web3.utils.utf8ToHex("UMATEST"), liquidationTime, disputePrice);
+
+    // Withdraw from liquidation to settle the dispute event.
+    const txObject2 = await emp.withdrawLiquidation("0", sponsor2, { from: sponsor2 });
+    await eventClient.clearState();
+
+    // Update the eventClient and check it has the dispute event stored correctly
+    await eventClient._update();
+
+    await contractMonitor.checkForNewDisputeSettlementEvents(() => toWei("1"));
+
+    assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/address/${sponsor1}`)); // liquidator address
+    assert.isFalse(lastSpyLogIncludes(spy, "(Monitored liquidator bot)")); // This liquidator is not monitored
+    assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/address/${sponsor2}`)); // disputer address
+    assert.isFalse(lastSpyLogIncludes(spy, "(Monitored dispute bot)")); // This disputer is not monitored
+    assert.isTrue(lastSpyLogIncludes(spy, "success")); // the disputed was successful based on settlement price
+    assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/tx/${txObject2.tx}`));
```

Comment:
These tests are awesome and easy to understand! Are you expecting the output to be in a certain Markdown format? 
Another way you could better modularize these tests and make the tests smaller would be to add a "testLog" (or something named like that) method to the module you're actually testing, `ContractMonitor` in this case.

The `testLog` method could take in a `spy` object and verify whether it is formatted correctly. Then the downstream unit test could just call `testLog(spy)`.

I'm not strongly opinionated, but it could aid test-driven-development by making the ContractMonitor developer more aware of how their logs should be structured.

Reply:
So I specifically tried to not validate format as this would couple the tests to a specific format and if the formatting was changed, even a small amount, all the tests would start failing. The reason these tests are set up like this it to be independent of the formatting and only validate this key information.